File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/SignCmd.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import picocli.CommandLine;
import sop.MicAlg;
import sop.ReadyWithResult;
import sop.SigningResult;
import sop.cli.picocli.SopCLI;
import sop.enums.SignAs;
import sop.exception.SOPGPException;
import sop.operation.DetachedSign;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

@CommandLine.Command(name = "sign",
        resourceBundle = "sop",
        exitCodeOnInvalidInput = 37)
public class SignCmd extends AbstractSopCmd {

    @CommandLine.Option(names = "--no-armor",
            descriptionKey = "sop.sign.usage.option.armor",
            negatable = true)
    boolean armor = true;

    @CommandLine.Option(names = "--as",
            descriptionKey = "sop.sign.usage.option.as",
            paramLabel = "{binary|text}")
    SignAs type;

    @CommandLine.Parameters(descriptionKey = "sop.sign.usage.parameter.keys",
            paramLabel = "KEYS")
    List<String> secretKeyFile = new ArrayList<>();

    @CommandLine.Option(names = "--with-key-password",
            descriptionKey = "sop.sign.usage.option.with_key_password",
            paramLabel = "PASSWORD")
    List<String> withKeyPassword = new ArrayList<>();

    @CommandLine.Option(names = "--micalg-out",
            descriptionKey = "sop.sign.usage.option.micalg_out",
            paramLabel = "MICALG")
    String micAlgOut;

    @Override
    public void run() {
        DetachedSign detachedSign = throwIfUnsupportedSubcommand(
                SopCLI.getSop().detachedSign(), "sign");

        throwIfOutputExists(micAlgOut);
        throwIfEmptyParameters(secretKeyFile, "KEYS");

        if (type != null) {
            try {
                detachedSign.mode(type);
            } catch (SOPGPException.UnsupportedOption unsupportedOption) {
                String errorMsg = getMsg("sop.error.feature_support.option_not_supported", "--as");
                throw new SOPGPException.UnsupportedOption(errorMsg, unsupportedOption);
            }
        }

        for (String passwordFile : withKeyPassword) {
            try {
                String password = stringFromInputStream(getInput(passwordFile));
                detachedSign.withKeyPassword(password);
            } catch (SOPGPException.UnsupportedOption unsupportedOption) {
                String errorMsg = getMsg("sop.error.feature_support.option_not_supported", "--with-key-password");
                throw new SOPGPException.UnsupportedOption(errorMsg, unsupportedOption);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        for (String keyInput : secretKeyFile) {
            try (InputStream keyIn = getInput(keyInput)) {
                detachedSign.key(keyIn);
            } catch (IOException e) {
                throw new RuntimeException(e);
            } catch (SOPGPException.KeyIsProtected keyIsProtected) {
                String errorMsg = getMsg("sop.error.runtime.cannot_unlock_key", keyInput);
                throw new SOPGPException.KeyIsProtected(errorMsg, keyIsProtected);
            } catch (SOPGPException.BadData badData) {
                String errorMsg = getMsg("sop.error.input.not_a_private_key", keyInput);
                throw new SOPGPException.BadData(errorMsg, badData);
            }
        }

        if (!armor) {
            detachedSign.noArmor();
        }

        try {
            ReadyWithResult<SigningResult> ready = detachedSign.data(System.in);
            SigningResult result = ready.writeTo(System.out);

            MicAlg micAlg = result.getMicAlg();
            if (micAlgOut != null) {
                // Write micalg out
                OutputStream outputStream = getOutput(micAlgOut);
                micAlg.writeTo(outputStream);
                outputStream.close();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/SignCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import java.io.IOException
import picocli.CommandLine.*
import sop.cli.picocli.SopCLI
import sop.enums.SignAs
import sop.exception.SOPGPException
import sop.exception.SOPGPException.BadData
import sop.exception.SOPGPException.KeyIsProtected

@Command(
    name = "sign",
    resourceBundle = "msg_detached-sign",
    exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
class SignCmd : AbstractSopCmd() {

    @Option(names = [OPT_NO_ARMOR], negatable = true) var armor: Boolean = true

    @Option(names = [OPT_AS], paramLabel = "{binary|text}") var type: SignAs? = null

    @Parameters(paramLabel = "KEYS") var secretKeyFile: List<String> = listOf()

    @Option(names = [OPT_WITH_KEY_PASSWORD], paramLabel = "PASSWORD")
    var withKeyPassword: List<String> = listOf()

    @Option(names = [OPT_MICALG_OUT], paramLabel = "MICALG") var micAlgOut: String? = null

    override fun run() {
        val detachedSign = throwIfUnsupportedSubcommand(SopCLI.getSop().detachedSign(), "sign")

        throwIfOutputExists(micAlgOut)
        throwIfEmptyParameters(secretKeyFile, "KEYS")

        type?.let { throwIfUnsupportedOption(OPT_AS) { detachedSign.mode(it) } }

        withKeyPassword.forEach { passIn ->
            try {
                throwIfUnsupportedOption(OPT_WITH_KEY_PASSWORD) {
                    val password = stringFromInputStream(getInput(passIn))
                    detachedSign.withKeyPassword(password)
                }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }

        secretKeyFile.forEach { keyIn ->
            try {
                getInput(keyIn).use { input -> detachedSign.key(input) }
            } catch (ioe: IOException) {
                throw RuntimeException(ioe)
            } catch (keyIsProtected: KeyIsProtected) {
                val errorMsg = getMsg("sop.error.runtime.cannot_unlock_key", keyIn)
                throw KeyIsProtected(errorMsg, keyIsProtected)
            } catch (badData: BadData) {
                val errorMsg = getMsg("sop.error.input.not_a_private_key", keyIn)
                throw BadData(errorMsg, badData)
            }
        }

        if (!armor) {
            throwIfUnsupportedOption(OPT_NO_ARMOR) { detachedSign.noArmor() }
        }

        try {
            val ready = detachedSign.data(System.`in`)
            val result = ready.writeTo(System.out)

            if (micAlgOut != null) {
                getOutput(micAlgOut).use { result.micAlg.writeTo(it) }
            }
        } catch (e: IOException) {
            throw java.lang.RuntimeException(e)
        }
    }

    companion object {
        const val OPT_NO_ARMOR = "--no-armor"
        const val OPT_AS = "--as"
        const val OPT_WITH_KEY_PASSWORD = "--with-key-password"
        const val OPT_MICALG_OUT = "--micalg-out"
    }
}
